src/checkers/inference/VariableAnnotator.java

```diff
@@ -255,7 +255,7 @@ private VariableSlot createVariable(final Tree tree) {
      *                actual implied tree appended to it.
      * @return A new VariableSlot corresponding to tree
      */
-    private VariableSlot createVariable(final AnnotationLocation location) {
+    protected VariableSlot createVariable(final AnnotationLocation location) {
```

Comment:
Why make this method `protected`? Do you use this method outside of `VariableAnnotator`? If so, what is your use case?

Reply:
This is due to PICOInference. In PICOInference, I override this method to also generate inequality constraint to let every `VariableSlot` not equal to `@Bottom`

Reply:
There was already such method `createVariable()` that only calls `slotManager.createVariableSlot()` method. I found that in PICOInference, if I override this method to have some additional logic, I don't need to change other places to achieve my goal. This method gives us way to have additional business logic that `VariableAnnotator` can have.

Reply:
@topnessman Mier, the method name `createVariable()` doesn't implies it may create additional constraints. Even you override this within your own PICO system, it still require changing the infrastructure for this special purpose, and also your code in PICO may not be very easy to understand... (i.e. why only `VarSlot` created by this `createVariable()` method would constrained to not equal to `@Bottom`? This might be a silly question but for a new comer he/she may confuse.)

Regarding your goal, if it was " let every VariableSlot not equal to `@Bottom`", then could you do a pass in encoding phase somewhere to generate inequality constraints for every VariableAnno passed in?

I personally don't think it is a good idea to extend `VariableAnnotator` for only adding these simple inequality constraints....

Reply:
I kind of agree that generating inequality constraint when overriding `createVariable()` method is not that good. But here is my reasoning for keep this:
For normal `VariableSlot`(not `CombVariableSlot` or `ExistentialVariableSlot`), i.e. for slots that will be inserted back to the source code, PICO doesn't allow inferring to `@Bottom`. `VariableAnnotator` currently has this API - `createVariable()` to apply for a new `VariableSlot` from `slotManager`.
If we do this in lower level, say during the encoding phase, we can't print the inequality constraint from the debug solver. Because we humans are not sensitive to numbers(if you print clauses in MaxSAT backend), you can't easily know what a certain clause/solver encoding type for different backend mean. This happened before in GUT - I forgot that I added logic of letting result not to be inferred to `@Bottom`.
Also, we need a second pass over all the `VariableSlot`s to generate such `Constraint` if we do it right after all constraints are collected, which is a waste of time.
I think there must be a reason why there exists such `createVariable()` method instead of directly calling `slotManager.createVariableSlot()`.
@jyluo also have some need to extend `VariableAnnotator`. PICOInfer also has some other reasons for extending `VariableAnnotator`.
So, overriding `createVariable()` isn't the most understandable way, but it works perfect for `@Bottom` inequality. We can think of a better way in the future PR.

Reply:
I agree more with Charles's point of view. If your type system requires that every slot never be bottom, then change the encoding of a slot. If that encoding doesn't apply to CombVarSlot or ExistentialVarSlot then make those more special. Generating less CNF clauses overall (eg 1 slot clause vs 1 slot clause + 1 constraint clause) will also make inference execute faster, or less likely to reach system memory maximums when running on very large programs.

The biggest headache and mess of the front end of the inference framework right now is that we generate slots and constraints in very unclear ways, sometimes in qual hierarchy, sometimes in tree annotators, sometimes in variable annotators. I don't think it is wise to have constraints generated within slot creation. This change seems to add more technical debt to this overall problem.

I think you should just directly call slotManager in higher level code that's calling variableannotator. Pack the logic into a util class if you must to make code maintenance easier.

Reply:
> change the encoding of a slot. If that encoding doesn't apply to CombVarSlot or ExistentialVarSlot then make those more special

 Can you clarify a bit?

> Generating less CNF clauses overall (eg 1 slot clause vs 1 slot clause + 1 constraint clause) will also make inference execute faster, or less likely to reach system memory maximums when running on very large programs.

From that number of variables and clauses, I think it has the same number of clauses. And there is no "slot clause". Slot is only the variable. No matter you generate constraints or not, you have to tell solver that please don't infer this var id to `@Bottom`, so you always need one clause to represent this. It's not related to performance I think, in the solving phase. But I do think in higher level, generating lots of `InequalityConstraint` WILL slot down inference. But is it really a good thing that the result you see(set of constraints) is not consistent with what is encoded into solver? 

> The biggest headache and mess of the front end of the inference framework right now is that we generate slots and constraints in very unclear ways, sometimes in qual hierarchy, sometimes in tree annotators, sometimes in variable annotators. I don't think it is wise to have constraints generated within slot creation. This change seems to add more technical debt to this overall problem.

Agree, it's so messy to generate constraints in different places.

> I think you should just directly call slotManager in higher level code that's calling variableannotator. Pack the logic into a util class if you must to make code maintenance easier.

Client of `variableannotator` is `ListTreeAnnotator`:
```java
 @Override
    public Void defaultAction(Tree node, AnnotatedTypeMirror type) {
        for (TreeAnnotator annotator : annotators) {
            annotator.visit(node, type);
        }

        return null;
    }
```
Can you give me an example here for your point?

Reply:
in classic max sat, every slot is encoded via 1 hot encoding no? if so, there must be a CNF clause to capture the 1 hot encoding (aka a well-formness constraint that there isn't 2+ qualifiers assigned to the same slot), hence a slot clause. To never have bottom in this, you can change the 1 hot to not have a bool variable for bottom at all, so that the slot itself can never take on bottom.

My point of view is, you should change your specific version of slot encoding so that you capture your wellformness constraint of no-bottom, rather than adding constraints during variable creation. Your earlier words seem to imply that CombVarSlot and ExistentialVarSlot can have bottom, so encode those two slots with the possibility of bottom.

> is it really a good thing that the result you see(set of constraints) is not consistent with what is encoded into solver?

How can it be inconsistent? Your slots are implicitly never bottom if you encode the slot such that it cannot possibly take on the value of bottom. If the debug output shows some InequalityConstraint then it's actually between two slots, a far more interesting output than a bunch of X cannot be Bottom spam.

> Client of variableannotator is ListTreeAnnotator

The technical dept must be paid at some point, and the whole variableannotator class should disappear to reduce the number of sources where slots and constraints are being generated.

Reply:
If we change slot encoding, that's backend specific. If I want to add a new backend, I need to do similar thing in the new backend. Is this right?

Reply:
Encoding is always relative to some backend. In my case, the generic Z3IntFormatTranslator backend I created (coming in some future PR) by default doesn't provide any kind of slot encoding, but my UnitsZ3IntFormatTranslator does declare a specific slot encoding for the units type system.

I would expect as a design consistency that if some type system were to support multiple (and possibly very different) backends, it would declare how its slots and constraints are encoded for each backend.

I get that your explain feature pairs clauses with constraints so you want constraints for explanations. The better architecture is to associate each slot with a wellformness constraint, so that when we encode a slot we in effect encode the wellformness constraint (eg 1-hot wellformness) and forward that encoded clause to the appropriate places in CFI. That way, if the wellformness constraint on the slot is the cause of unsat then you get a clearer explanation (and thus implemented as default behavior here in a updated VariableAnnotator). I would see this being useful across all type systems and not just useful to PICO; each type system can optionally implement its own wellformness constraint that if unsat has an explanation.

Reply:
What I'm concerning is that, we need to do the repetitive job in your supported backends to restrict the result not to be `@Bottom`. But if you generate `InequalityConstraint`, you only generate it without the need of doing the repetitive thing when you add a new backend.

> I get that your explain feature pairs clauses with constraints so you want constraints for explanations. The better architecture is to associate each slot with a wellformness constraint, so that when we encode a slot we in effect encode the wellformness constraint (eg 1-hot wellformness) and forward that encoded clause to the appropriate places in CFI. That way, if the wellformness constraint on the slot is the cause of unsat then you get a clearer explanation (and thus implemented as default behavior here in a updated VariableAnnotator). I would see this being useful across all type systems and not just useful to PICO; each type system can optionally implement its own wellformness constraint that if unsat has an explanation.

I think this needs us to change the solver framework furthur, am I right? For example, right now `serialize(XXXSlot)` returns null instead of generating well-formedness Constraint. Right now, our CFI doesn't have a subtype of `Constraint` to represent "well-formedness Constraint", how can we represent this in human-readable way rather than solver encoding result in your proposal?

If well-formedness Constraint is part of unsat reason, this PR only prints the clause because we don't have a type to represent it.

Reply:
Restricting the result is also high level semantics for type system. If we mix this logic into solver, it sounds like a hack for me.

Reply:
>if you generate InequalityConstraint

the total clauses isn't minimal for performance or scaling if you generate a 1-hot over all qualifiers and then a clause saying no bottom.

from my point of view, encode slot == encode wellformness constraint, which by default returns no clauses. for maxsat it returns the 1-hot encoding, for PICO you can directly return 1-hot over all your qualifiers except bottom.

>serialize(XXXSlot) returns null
In my units type system it doesn't return null, instead this is where I currently generate my wellformness clause (which is a mix of 1-hot over 2 booleans representing top and bottom, and also the set of z3 int variables; a complex 1-hot to be succinct)

>Constraint to represent "well-formedness Constraint"

We make a new top level subclass / generic implementation of a WellformnessConstraint

under this architecture, you subclass PICOSlotWellformnessConstraint and define your encoding of how a slot should never be bottom

>If well-formedness Constraint is part of unsat reason, this PR only prints the clause because we don't have a type to represent it.

1) wouldn't it be more clear as an output message to print firstly the constraint that's unsat and then the clauses for that constraint?

2) we can put in class WellformnessConstraint a reference to the slot ID, then we know which slot was unsat and then we can print that out (and even better, print all constraints associated with that slot in future PR).

No matter what my idea is a future PR task. I think that overriding createVariable() is not quite the right solution because to me each slot has a wellformness constraint baked in, be it 1-hot, no-bottom, or unit's complex 1-hot, and the wellformness constraints should become more visible in the system.

Reply:
> Restricting the result not to be @Bottom is also high level semantics/business logic for type system. If we mix this logic into solver, it sounds like a hack for me.

So I see this as your wellformenss constraint, which to me architecturally is best if it is closely attached to the code for encoding a slot.

Lets say you choose z3 (bool theory) and sat4j backends, you still need to provide two different encodings via 2 different type system specific format translators. I don't see how this is a hack.